From: Zhao Yakui <yakui.zhao@intel.com>
Subject: ACPI: Rename ACPI processor device bus ID
References: bnc#528769
Patch-Mainline: yes
Commit-ID: 7a04b8491a077471a34938b8ca060c37220953be

Signed-off-by: Thomas Renninger <trenn@suse.de>

Some BIOS re-use the same processor bus id
in different scope:

	\_SB.SCK0.CPU0
	\_SB.SCK1.CPU0

But the (deprecated) /proc/acpi/ interface
assumes the bus-id's are unique, resulting in an OOPS
when the processor driver is loaded:

WARNING: at fs/proc/generic.c:590 proc_register+0x148/0x180()
Hardware name: Sunrise Ridge
proc_dir_entry 'processor/CPU0' already registered
Call Trace:
 [<ffffffff8023f7ef>] warn_slowpath+0xb1/0xe5
 [<ffffffff8036243b>] ? ida_get_new_above+0x190/0x1b1
 [<ffffffff803625a8>] ? idr_pre_get+0x5f/0x75
 [<ffffffff8030b2f6>] proc_register+0x148/0x180
 [<ffffffff8030b4ff>] proc_mkdir_mode+0x3d/0x52
 [<ffffffff8030b525>] proc_mkdir+0x11/0x13
 [<ffffffffa0014b89>] acpi_processor_start+0x755/0x9bc [processor]

Rename the processor device bus id. And the new bus id will be
generated as the following format:
	CPU+ CPU ID

For example: If the cpu ID is 5, then the bus ID will be "CPU5".
	If the CPU ID is 10, then the bus ID will be "CPUA".

Yes, this will change the directory names seen
in /proc/acpi/processor/* on some systems.
Before this patch, those directory names where
totally arbitrary strings based on the interal AML device strings.

http://bugzilla.kernel.org/show_bug.cgi?id=13612

Signed-off-by: Zhao Yakui <yakui.zhao@intel.com>
Signed-off-by: Len Brown <len.brown@intel.com>

Automatically created from "patches.fixes/acpi_assure_unique_processor_proc_creation.patch" by xen-port-patches.py

--- sle11-2009-09-18.orig/drivers/acpi/processor_core.c	2009-09-24 11:28:54.000000000 +0200
+++ sle11-2009-09-18/drivers/acpi/processor_core.c	2009-09-21 09:15:34.000000000 +0200
@@ -634,7 +634,11 @@ static int acpi_processor_get_info(struc
 	 * generated as the following format:
 	 * CPU+CPU ID.
 	 */
-	sprintf(acpi_device_bid(device), "CPU%X", pr->id);
+	if (pr->id != -1)
+		sprintf(acpi_device_bid(device), "CPU%X", pr->id);
+	else
+		snprintf(acpi_device_bid(device), ACPI_DEVICE_BID_SIZE,
+			 "#%0*X", ACPI_DEVICE_BID_SIZE - 2, pr->acpi_id);
 	ACPI_DEBUG_PRINT((ACPI_DB_INFO, "Processor [%d:%d]\n", pr->id,
 			  pr->acpi_id));
 
--- sle11-2009-09-18.orig/include/acpi/acpi_bus.h	2009-09-24 11:28:54.000000000 +0200
+++ sle11-2009-09-18/include/acpi/acpi_bus.h	2009-09-24 11:29:48.000000000 +0200
@@ -199,6 +199,10 @@ struct acpi_device_pnp {
 	acpi_device_class device_class;	/*        "          */
 };
 
+#define ACPI_DEVICE_BID_SIZE	\
+	((unsigned int)(offsetof(struct acpi_device_pnp, bus_address) - \
+			offsetof(struct acpi_device_pnp, bus_id)))
+
 #define acpi_device_bid(d)	((d)->pnp.bus_id)
 #define acpi_device_adr(d)	((d)->pnp.bus_address)
 #define acpi_device_hid(d)	((d)->pnp.hardware_id)
